File: packages/flags-browser/src/evaluator.ts
/**
 * Flag evaluation logic.
 *
 * Evaluates a single flag or all flags for a given user context,
 * returning the variation key(s) the user should see.
 */

import type {
  ConfigFlag,
  ProjectConfig,
  VariationAllocation,
  RuleType,
  SignaKitDecision,
  SignaKitDecisions,
  UserAttributes,
  VariableValue,
} from './types'
import { hashForTraffic, hashForVariation, hashForDefault } from './hasher'
import { matchesAudiences } from './audience-matcher'
import { BUCKET_SPACE } from './constants'

/**
 * Find the variation for a given bucket within allocation ranges.
 *
 * @param allocation - The variation allocation with ranges
 * @param bucket - The bucket number (0-9999)
 * @returns The variation key or null if no range matches
 */
function findVariationInRanges(allocation: VariationAllocation, bucket: number): string | null {
  for (const range of allocation.ranges) {
    // Range is inclusive of both start and end
    if (bucket >= range.start && bucket <= range.end) {
      return range.variation
    }
  }
  return null
}

export interface EvaluationResult {
  variationKey: string
  enabled: boolean
  ruleKey: string | null
  /** The rule type that produced this decision, or null for default/disabled paths. */
  ruleType: RuleType | null
  variables: Record<string, VariableValue>
}

/**
 * Resolve variable values for a variation by merging flag-level defaults
 * with any variation-specific overrides.
 */
function resolveVariables(flag: ConfigFlag, variationKey: string): Record<string, VariableValue> {
  if (!flag.variables || flag.variables.length === 0) return {}

  const variation = flag.variations.find((v) => v.key === variationKey)
  const overrides = variation?.variables ?? {}

  const resolved: Record<string, VariableValue> = {}
  for (const def of flag.variables) {
    const override = overrides[def.key]
    resolved[def.key] = override !== undefined ? override : def.defaultValue
  }
  return resolved
}

/**
 * Evaluate a single flag for a user.
 *
 * Algorithm:
 * 1. If archived, skip (return null - flag won't appear in results)
 * 2. If not running, return disabled decision
 * 3. Evaluate rules in order:
 *    a. Check allowlist - if userId matches, return that variation immediately
 *    b. Check audience match (any/all logic)
 *    c. Check traffic allocation (bucket < trafficPercentage * 100)
 *    d. Get variation from allocation ranges
 * 4. No rules match → use default allocation
 *
 * @param flag - The flag configuration
 * @param userId - The user's unique identifier
 * @param attributes - Optional user attributes for targeting
 * @returns The evaluation result or null if archived
 */
export function evaluateFlag(
  flag: ConfigFlag,
  userId: string,
  attributes?: UserAttributes
): EvaluationResult | null {
  // 1. If archived, skip this flag entirely
  if (flag.status === 'archived') {
    return null
  }

  // 2. If not running, return disabled decision with 'off' variation
  if (!flag.running) {
    return {
      variationKey: 'off',
      enabled: false,
      ruleKey: null,
      ruleType: null,
      variables: resolveVariables(flag, 'off'),
    }
  }

  // 3. Evaluate rules in order (first matching rule wins)
  const rules = flag.rules || []
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i]
    if (!rule) continue

    // 3a. Check allowlist first - if userId is in allowlist, return that variation immediately
    if (rule.allowlist && rule.allowlist.length > 0) {
      const allowlistEntry = rule.allowlist.find((entry) => entry.userId === userId)
      if (allowlistEntry) {
        return {
          variationKey: allowlistEntry.variation,
          enabled: allowlistEntry.variation !== 'off',
          ruleKey: rule.ruleKey,
          ruleType: rule.ruleType,
          variables: resolveVariables(flag, allowlistEntry.variation),
        }
      }
    }

    // 3b. Check if user matches the rule's audiences
    if (matchesAudiences(rule.audiences, rule.audienceMatchType, attributes)) {
      // Calculate traffic bucket for this rule
      const trafficBucket = hashForTraffic(flag.salt, userId)

      // trafficPercentage is 0-100, so multiply by 100 to get bucket threshold (0-10000)
      // e.g., 50% traffic = 5000 threshold (buckets 0-4999 are in traffic)
      const trafficThreshold = Math.floor((rule.trafficPercentage / 100) * BUCKET_SPACE)

      // If user is in traffic for this rule
      if (trafficBucket < trafficThreshold) {
        // Get variation bucket
        const variationBucket = hashForVariation(flag.salt, userId)
        const variation = findVariationInRanges(rule.variationAllocation, variationBucket)

        if (variation) {
          return {
            variationKey: variation,
            enabled: variation !== 'off',
            ruleKey: rule.ruleKey,
            ruleType: rule.ruleType,
            variables: resolveVariables(flag, variation),
          }
        }
      }
    }
  }

  // 4. No rules matched - use default allocation
  const defaultBucket = hashForDefault(flag.salt, userId)
  const defaultVariation = findVariationInRanges(flag.allocation, defaultBucket)

  if (defaultVariation) {
    return {
      variationKey: defaultVariation,
      enabled: defaultVariation !== 'off',
      ruleKey: null,
      ruleType: null,
      variables: resolveVariables(flag, defaultVariation),
    }
  }

  // Fallback to disabled if no allocation matched
  return {
    variationKey: 'off',
    enabled: false,
    ruleKey: null,
    ruleType: null,
    variables: resolveVariables(flag, 'off'),
  }
}

/**
 * Evaluate all flags in a config for a user.
 *
 * @param config - The project configuration with all flags
 * @param userId - The user's unique identifier
 * @param attributes - Optional user attributes for targeting
 * @returns Map of flag keys to decisions
 */
export function evaluateAllFlags(
  config: ProjectConfig,
  userId: string,
  attributes?: UserAttributes
): SignaKitDecisions {
  const decisions: SignaKitDecisions = {}

  for (const flag of config.flags) {
    const result = evaluateFlag(flag, userId, attributes)

    // Only include non-null results (excludes archived flags)
    if (result !== null) {
      const decision: SignaKitDecision = {
        flagKey: flag.key,
        variationKey: result.variationKey,
        enabled: result.enabled,
        ruleKey: result.ruleKey,
        ruleType: result.ruleType,
        variables: result.variables,
      }
      decisions[flag.key] = decision
    }
  }

  return decisions
}

File: packages/flags-java/src/main/java/com/signakit/flags/evaluator/Evaluator.java
package com.signakit.flags.evaluator;

import com.signakit.flags.Constants;
import com.signakit.flags.Decision;
import com.signakit.flags.audience.AudienceMatcher;
import com.signakit.flags.config.ConfigModels.AllowlistEntry;
import com.signakit.flags.config.ConfigModels.ConfigFlag;
import com.signakit.flags.config.ConfigModels.ConfigRule;
import com.signakit.flags.config.ConfigModels.FlagStatus;
import com.signakit.flags.config.ConfigModels.FlagVariable;
import com.signakit.flags.config.ConfigModels.ProjectConfig;
import com.signakit.flags.config.ConfigModels.Variation;
import com.signakit.flags.config.ConfigModels.VariationAllocation;
import com.signakit.flags.config.ConfigModels.VariationAllocationRange;
import com.signakit.flags.hasher.Hasher;

import java.util.LinkedHashMap;
import java.util.Map;

/** Evaluates a single flag (or all flags) for a given user. */
public final class Evaluator {
    private Evaluator() {}

    private static String findVariationInRanges(VariationAllocation allocation, int bucket) {
        if (allocation == null || allocation.ranges == null) return null;
        for (VariationAllocationRange r : allocation.ranges) {
            if (bucket >= r.start && bucket <= r.end) return r.variation;
        }
        return null;
    }

    private static Map<String, Object> resolveVariables(ConfigFlag flag, String variationKey) {
        Map<String, Object> out = new LinkedHashMap<>();
        if (flag.variables == null || flag.variables.isEmpty()) return out;
        Map<String, Object> overrides = Map.of();
        if (flag.variations != null) {
            for (Variation v : flag.variations) {
                if (variationKey.equals(v.key) && v.variables != null) {
                    overrides = v.variables;
                    break;
                }
            }
        }
        for (FlagVariable def : flag.variables) {
            Object override = overrides.get(def.key);
            out.put(def.key, override != null ? override : def.defaultValue);
        }
        return out;
    }

    /**
     * Evaluate a single flag. Returns {@code null} when the flag is archived.
     */
    public static Decision evaluateFlag(ConfigFlag flag, String userId, Map<String, Object> attributes) {
        if (flag.status == FlagStatus.ARCHIVED) return null;

        if (!flag.running) {
            return new Decision(
                    flag.key, "off", false, null, null, resolveVariables(flag, "off"));
        }

        if (flag.rules != null) {
            for (ConfigRule rule : flag.rules) {
                if (rule == null) continue;

                // Allowlist
                if (rule.allowlist != null && !rule.allowlist.isEmpty()) {
                    for (AllowlistEntry entry : rule.allowlist) {
                        if (userId.equals(entry.userId)) {
                            String variation = entry.variation;
                            return new Decision(
                                    flag.key,
                                    variation,
                                    !"off".equals(variation),
                                    rule.ruleKey,
                                    rule.ruleType,
                                    resolveVariables(flag, variation));
                        }
                    }
                }

                if (!AudienceMatcher.matchesAudiences(rule.audiences, rule.audienceMatchType, attributes)) {
                    continue;
                }

                int trafficBucket = Hasher.hashForTraffic(flag.salt, userId);
                int trafficThreshold = (int) Math.floor((rule.trafficPercentage / 100.0) * Constants.BUCKET_SPACE);
                if (trafficBucket >= trafficThreshold) continue;

                int variationBucket = Hasher.hashForVariation(flag.salt, userId);
                String variation = findVariationInRanges(rule.variationAllocation, variationBucket);
                if (variation != null) {
                    return new Decision(
                            flag.key,
                            variation,
                            !"off".equals(variation),
                            rule.ruleKey,
                            rule.ruleType,
                            resolveVariables(flag, variation));
                }
            }
        }

        // Default allocation
        int defaultBucket = Hasher.hashForDefault(flag.salt, userId);
        String defaultVariation = findVariationInRanges(flag.allocation, defaultBucket);
        if (defaultVariation != null) {
            return new Decision(
                    flag.key,
                    defaultVariation,
                    !"off".equals(defaultVariation),
                    null,
                    null,
                    resolveVariables(flag, defaultVariation));
        }

        return new Decision(flag.key, "off", false, null, null, resolveVariables(flag, "off"));
    }

    public static Map<String, Decision> evaluateAllFlags(
            ProjectConfig config, String userId, Map<String, Object> attributes) {
        Map<String, Decision> out = new LinkedHashMap<>();
        if (config.flags == null) return out;
        for (ConfigFlag flag : config.flags) {
            Decision d = evaluateFlag(flag, userId, attributes);
            if (d != null) out.put(flag.key, d);
        }
        return out;
    }
}
